### DISCUSSION TITLE
Proxy chain issue

### DISCUSSION THREAD

**Discussion by carlos-suarez** [mitmproxy/mitmproxy#7104] [General]
Hope this is not too confusing, will try to explain as simple as possible.

I'm invested already on a multi-year developed solution working for me using FiddlerClassic proxy (Windows) where a lot of code is already in a DLL for efficiency and flexibility.  A challenge came where I need to route traffic thru different IPs depending of the client connected to my FiddlerProxy but fiddler doesn't support binding to different IPs for outgoing requests so I came with this quick kind of dirty solution to forward the traffic depending on the client criteria to small instances of mitmproxy running on Linux (Ubuntu).  It works fine and was an easy setup for the purpose of this question I'm going simplify like this:

FiddlerProxy listens on 198.0.0.1:8080
mitmproxy1 listens on 198.0.0.2:8080
mitmproxy2 listens on 198.0.0.3:8080

Clients connects via proxy to 198.0.0.1:8080 and HTTPS request gets captured and inspected in FiddlerProxy.  Then based in some criteria I use the following in 'OnBeforeRequest':

if (true)
   mitmproxy = "198.0.0.2";
else
   mitmproxy = "198.0.0.3";

oSession["x-overrideHost"] = mitmproxy + ":7887"; 
oSession["x-overrideGateway"] = "DIRECT";

as I said that works perfect.  Now here is the kick, I did this with Linux servers running on AWS (Amazon), then I had the grate idea of diversifying to lower my risk and started creating GCP instances.. all happy happy then I went for my next cloud provider Azure and oh for my big surprise those instances don't want to play nice LOL.

I'm using mimtdump but for POC purpose I was able to reproduce the same with mimtproxy and this is all in clean out of the box Ubuntu running only 3 commands:

sudo apt update
sudo apt upgrade
sudo apt install mitm

then running mimtproxy like:

mitmproxy --set listen_port=8080 --set block_global=false

AWS, GCP instances work fine but Azure instance give the error:

"Cannot connect to server, no server address given"

and I'm totally lost.

Just for kicks I create instances on other 2 different providers Linode and Kamatera, both work.. and a third one LiquidWeb I get the same error like Azure.

Worth to mention if I connect the clients directly to those instances mitmproxy works fine, this is only when I try to re-direct going thru my FiddlerProxy instance.

Hope someone can help :-)

**Comment by mhils** [comment]
Have you looked at `mitmdump --version` on the different machines?

**Reply by carlos-suarez** [reply]
Good point, never thought of checking that since it was working so easily on AWS and GCP and I was supposedly doing sudo apt upgrade before installing mitmproxy lol..

On the working instance (AWS) I get this:
Mitmproxy: 8.1.1
Python:    3.12.3
OpenSSL:   OpenSSL 3.0.13 30 Jan 2024
Platform:  Linux-6.8.0-1012-aws-x86_64-with-glibc2.39

On the non-working (LiquidWeb) I get:
Mitmproxy: 6.0.2
Python:    3.10.12
OpenSSL:   OpenSSL 3.0.2 15 Mar 2022
Platform:  Linux-5.15.0-118-generic-x86_64-with-glibc2.35

So I guess now I need to try to upgrade somehow right?  .. not linux expert but will give it a try :-)

Or is this that I have to point my sudo apt install to some repo so it installs the latest version?

**Reply by carlos-suarez** [reply]
And works like a charm now of course! .. thank you!!!!

Just to mention what I did:
sudo apt remove mitmproxy - to remove old mitmproxy
sudo apt install python3-pip - for alternate installation method
sudo pip3 install mitmproxy - to get latest mitmproxy, got version 10

and now it works! :-)

**Reply by mhils** [reply]
Happy to hear this helped! For reference, here are our recommended installation methods: https://docs.mitmproxy.org/stable/overview-installation/

**Reply by carlos-suarez** [reply]
Thank you that was really helpful.  I guess I'll be more diligent on checking the versions next time :-).  I'm more of Window guy, the little linux I do is totally google guided hehe.

One more question if I might.  I've been looking at the options specifically the use of client and server ciphers as a way to bypass TLS fingerprinting.  I have this app that I'm analyzing the traffic for which is blocking with 403, after some reading it seems that's the reason and as king of experimentation (or mostly luck), I found the some lower version of mimtproxy do not get blocked while higher versions do.  I don't see however difference in the ciphers and using those options do not seems to have a real effect but then I'm not expert on that either and maybe I'm not doing or looking at it correctly.  My understanding is I need to focus on **ciphers_server** because that's the one controlling **mitmproxy <-> server connections**, is that correct? any additional considerations or guidance I'd really appreciate.  Maybe I need to make this a new conversation all together?

**Reply by mhils** [reply]
https://github.com/mitmproxy/mitmproxy/issues/4575

**Reply by carlos-suarez** [reply]
Man this is all way beyond my pay grade now lol.. I'll see where I can take it from here, at least now with version 7 is working for me but I suppose it'll be just a matter of time till I get blocked again :-\.  Thanks for the help tho :-)
